src/server/worker/driver/driver.go

```diff
@@ -0,0 +1,1233 @@
+	}
+
+	// Clone checks out a reference, not a SHA. Github does not support fetching
+	// an individual SHA.
+	remoteURL := payload.Repository.CloneURL
+	gitRepo, err := git.PlainCloneContext(
+		d.pachClient.Ctx(),
+		filepath.Join(scratchPath, input.Name),
+		false,
+		&git.CloneOptions{
+			URL:           remoteURL,
+			SingleBranch:  true,
+			ReferenceName: gitPlumbing.ReferenceName(payload.Ref),
+		},
+	)
+	if err != nil {
+		return errors.Wrapf(err, "error fetching repo %v with ref %v from URL %v", input.Name, payload.Ref, remoteURL)
+	}
+
+	wt, err := gitRepo.Worktree()
+	if err != nil {
+		return err
+	}
+
+	sha := payload.After
+	err = wt.Checkout(&git.CheckoutOptions{Hash: gitPlumbing.NewHash(sha)})
+	if err != nil {
+		return errors.Wrapf(err, "error checking out SHA %v for repo %v", sha, input.Name)
+	}
+
+	// go-git will silently fail to checkout an invalid SHA and leave the HEAD at
+	// the selected ref. Verify that we are now on the correct SHA
+	rev, err := gitRepo.ResolveRevision("HEAD")
+	if err != nil {
+		return errors.Wrapf(err, "failed to inspect HEAD SHA for repo %v", input.Name)
+	} else if rev.String() != sha {
+		return errors.Errorf("could not find SHA %v for repo %v", sha, input.Name)
+	}
+
+	return nil
+}
+
+// Run user code and return the combined output of stdout and stderr.
+func (d *driver) RunUserCode(
+	logger logs.TaggedLogger,
+	environ []string,
+	procStats *pps.ProcessStats,
+	rawDatumTimeout *types.Duration,
+) (retErr error) {
+	ctx := d.pachClient.Ctx()
+	d.reportUserCodeStats(logger)
+	defer func(start time.Time) { d.reportDeferredUserCodeStats(retErr, start, procStats, logger) }(time.Now())
+	logger.Logf("beginning to run user code")
+	defer func(start time.Time) {
+		if retErr != nil {
+			logger.Logf("errored running user code after %v: %v", time.Since(start), retErr)
+		} else {
+			logger.Logf("finished running user code after %v", time.Since(start))
+		}
+	}(time.Now())
+	if rawDatumTimeout != nil {
+		datumTimeout, err := types.DurationFromProto(rawDatumTimeout)
+		if err != nil {
+			return err
+		}
+		datumTimeoutCtx, cancel := context.WithTimeout(ctx, datumTimeout)
+		defer cancel()
+		ctx = datumTimeoutCtx
+	}
+
+	if len(d.pipelineInfo.Transform.Cmd) == 0 {
+		return errors.New("invalid pipeline transform, no command specified")
+	}
+
+	// Run user code
+	cmd := exec.CommandContext(ctx, d.pipelineInfo.Transform.Cmd[0], d.pipelineInfo.Transform.Cmd[1:]...)
+	if d.pipelineInfo.Transform.Stdin != nil {
+		cmd.Stdin = strings.NewReader(strings.Join(d.pipelineInfo.Transform.Stdin, "\n") + "\n")
+	}
+	cmd.Stdout = logger.WithUserCode()
+	cmd.Stderr = logger.WithUserCode()
+	cmd.Env = environ
+	if d.uid != nil && d.gid != nil {
+		cmd.SysProcAttr = makeCmdCredentials(*d.uid, *d.gid)
+	}
+	cmd.Dir = filepath.Join(d.rootDir, d.pipelineInfo.Transform.WorkingDir)
+	err := cmd.Start()
+	if err != nil {
+		return errors.Wrap(err, "cmd.Start")
```

Comment:
Do we want this to be an `errors.Ensure`? It's not totally clear to me, since these logs get passed to users and they may just look at the error string, it might be good to have the information duplicated in the err string and the stack.